File: frontend/src/api/decision.ts
/**
 * Phase 6 — API client for Decision Narrative & Human Award Workflow.
 */

export type NarrativeType =
  | "comparison_summary"
  | "tradeoff_analysis"
  | "sensitivity_summary"
  | "decision_support_memo"
  | "decision_considerations";

export type NarrativeOrigin = "ai_generated" | "ai_generated_human_revised" | "human_authored";
export type ClaimType = "deterministic_fact" | "interpretation" | "limitation";
export type GroundingStatus = "verified" | "unsupported" | "unverifiable";
export type AwardStatus = "draft" | "confirmed" | "revoked";
export type AwardEventType = "draft_created" | "confirmed" | "revoked";

export interface FactReference {
  reference_type: string;
  field_path?: string;
  authoritative_value?: string;
  criterion_id?: string;
  supplier_id?: string;
  evidence_id?: string;
}

export interface NarrativeClaim {
  id: string;
  claim_index: number;
  text: string;
  claim_type: ClaimType;
  grounding_status: GroundingStatus;
  referenced_supplier_ids: string[];
  referenced_criterion_ids: string[];
  referenced_evidence_ids: string[];
  fact_references?: FactReference[] | Record<string, any> | null;
  grounding_notes?: string | null;
}

export interface NarrativeRevision {
  id: string;
  narrative_generation_id: string;
  revision_number: number;
  revised_text: string;
  revision_rationale?: string | null;
  revised_by: string;
  created_at: string;
}

export interface SupplierAnalysis {
  supplier_id: string;
  supplier_name: string;
  rank?: number | null;
  total_score: string;
  strengths: string[];
  weaknesses: string[];
  score_context: string;
}

export interface NarrativeSections {
  executive_summary: string;
  ranking_explanation: string;
  per_supplier_analysis: SupplierAnalysis[];
  trade_offs: string;
  decision_considerations: string;
  risk_factors: string[];
  data_limitations: string[];
}

export interface GroundingValidationResult {
  total_claims: number;
  verified: number;
  unsupported: number;
  unverifiable: number;
  details: Array<{
    claim_index: number;
    status: string;
    issues: string[];
  }>;
}

export interface NarrativeGenerationResponse {
  id: string;
  rfq_id: string;
  decision_context_id: string;
  narrative_type: NarrativeType;
  generation_number: number;
  origin: NarrativeOrigin;
  provider: string;
  model_identifier: string;
  prompt_template_version: string;
  prompt_template_hash: string;
  rendered_prompt_hash: string;
  response_schema_version: string;
  generation_parameters: Record<string, any>;
  raw_structured_output: NarrativeSections;
  output_hash: string;
  grounding_validation_result: GroundingValidationResult;
  is_superseded: boolean;
  superseded_reason?: string | null;
  generated_at: string;
  created_by: string;
  claims: NarrativeClaim[];
  revisions: NarrativeRevision[];
  current_origin: NarrativeOrigin;
}

export interface NarrativeGenerationRequest {
  scoring_run_id: string;
  narrative_type?: NarrativeType;
  include_sensitivity?: boolean;
  human_unverified_note?: string | null;
}

export interface DecisionContextResponse {
  id: string;
  rfq_id: string;
  scoring_run_id: string;
  context_schema_version: string;
  context_payload: Record<string, any>;
  context_hash: string;
  includes_sensitivity: boolean;
  created_by: string;
  created_at: string;
}

export interface AwardDecisionEvent {
  id: string;
  award_decision_id: string;
  event_type: AwardEventType;
  event_number: number;
  event_payload: Record<string, any>;
  actor_principal: string;
  actor_display_name?: string | null;
  created_at: string;
}

export interface AwardDecisionResponse {
  id: string;
  rfq_id: string;
  scoring_run_id: string;
  narrative_generation_id?: string | null;
  awarded_supplier_id: string;
  awarded_supplier_name: string;
  awarded_supplier_rank?: number | null;
  non_rank1_rationale?: string | null;
  current_status: AwardStatus;
  provenance_hash?: string | null;
  created_by: string;
  created_at: string;
  events: AwardDecisionEvent[];
  is_based_on_latest_run: boolean;
  latest_scoring_run_id?: string | null;
}

export interface AwardDecisionCreate {
  scoring_run_id: string;
  awarded_supplier_id: string;
  award_justification: string;
  narrative_generation_id?: string | null;
  non_rank1_rationale?: string | null;
}

export interface AwardConfirm {
  final_justification?: string | null;
}

export interface AwardRevoke {
  revocation_reason: string;
}

const API_BASE = import.meta.env.VITE_API_BASE_URL || import.meta.env.VITE_API_URL || "";
const API_KEY = "procureflow_dev_api_key_12345";

export async function generateNarrative(
  rfqId: string,
  data: NarrativeGenerationRequest
): Promise<NarrativeGenerationResponse> {
  const res = await fetch(`${API_BASE}/api/v1/rfqs/${rfqId}/decisions/narratives`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-API-Key": API_KEY,
    },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || "Failed to generate narrative");
  }
  return res.json();
}

export async function listNarratives(
  rfqId: string,
  scoringRunId?: string
): Promise<NarrativeGenerationResponse[]> {
  const query = scoringRunId ? `?scoring_run_id=${encodeURIComponent(scoringRunId)}` : "";
  const res = await fetch(`${API_BASE}/api/v1/rfqs/${rfqId}/decisions/narratives${query}`, {
    headers: { "X-API-Key": API_KEY },
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || "Failed to list narratives");
  }
  return res.json();
}

export async function getNarrative(
  rfqId: string,
  narrativeId: string
): Promise<NarrativeGenerationResponse> {
  const res = await fetch(
    `${API_BASE}/api/v1/rfqs/${rfqId}/decisions/narratives/${narrativeId}`,
    {
      headers: { "X-API-Key": API_KEY },
    }
  );
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || "Failed to get narrative");
  }
  return res.json();
}

export async function createNarrativeRevision(
  rfqId: string,
  narrativeId: string,
  revisedText: string,
  rationale?: string
): Promise<NarrativeRevision> {
  const res = await fetch(
    `${API_BASE}/api/v1/rfqs/${rfqId}/decisions/narratives/${narrativeId}/revisions`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-API-Key": API_KEY,
      },
      body: JSON.stringify({
        revised_text: revisedText,
        revision_rationale: rationale,
      }),
    }
  );
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || "Failed to record narrative revision");
  }
  return res.json();
}

export async function getDecisionContext(
  rfqId: string,
  contextId: string
): Promise<DecisionContextResponse> {
  const res = await fetch(
    `${API_BASE}/api/v1/rfqs/${rfqId}/decisions/contexts/${contextId}`,
    {
      headers: { "X-API-Key": API_KEY },
    }
  );
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || "Failed to get decision context");
  }
  return res.json();
}

export async function createDraftAward(
  rfqId: string,
  data: AwardDecisionCreate
): Promise<AwardDecisionResponse> {
  const res = await fetch(`${API_BASE}/api/v1/rfqs/${rfqId}/decisions/awards`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-API-Key": API_KEY,
    },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || "Failed to create draft award");
  }
  return res.json();
}

export async function listAwards(rfqId: string): Promise<AwardDecisionResponse[]> {
  const res = await fetch(`${API_BASE}/api/v1/rfqs/${rfqId}/decisions/awards`, {
    headers: { "X-API-Key": API_KEY },
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || "Failed to list awards");
  }
  return res.json();
}

export async function getCurrentAward(
  rfqId: string
): Promise<AwardDecisionResponse | null> {
  const res = await fetch(
    `${API_BASE}/api/v1/rfqs/${rfqId}/decisions/awards/current`,
    {
      headers: { "X-API-Key": API_KEY },
    }
  );
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || "Failed to get current award");
  }
  return res.json();
}

export async function confirmAward(
  rfqId: string,
  awardId: string,
  data: AwardConfirm
): Promise<AwardDecisionResponse> {
  const res = await fetch(
    `${API_BASE}/api/v1/rfqs/${rfqId}/decisions/awards/${awardId}/confirm`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-API-Key": API_KEY,
      },
      body: JSON.stringify(data),
    }
  );
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || "Failed to confirm award");
  }
  return res.json();
}

export async function revokeAward(
  rfqId: string,
  awardId: string,
  data: AwardRevoke
): Promise<AwardDecisionResponse> {
  const res = await fetch(
    `${API_BASE}/api/v1/rfqs/${rfqId}/decisions/awards/${awardId}/revoke`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-API-Key": API_KEY,
      },
      body: JSON.stringify(data),
    }
  );
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || "Failed to revoke award");
  }
  return res.json();
}

File: backend/src/procureflow/schemas/decision.py
"""
Phase 6 — Pydantic schemas for Decision Narrative & Human Award Workflow.

Covers narrative generation requests/responses, structured claims with
grounding validation, award lifecycle events, and DecisionContext.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator

from procureflow.utils.actor import sanitize_actor_string as _sanitize_actor_string

# ---------------------------------------------------------------------------
# ENUMERATIONS
# ---------------------------------------------------------------------------


class NarrativeType(str, Enum):
    COMPARISON_SUMMARY = "comparison_summary"
    TRADEOFF_ANALYSIS = "tradeoff_analysis"
    SENSITIVITY_SUMMARY = "sensitivity_summary"
    DECISION_SUPPORT_MEMO = "decision_support_memo"
    DECISION_CONSIDERATIONS = "decision_considerations"


class NarrativeOrigin(str, Enum):
    AI_GENERATED = "ai_generated"
    AI_GENERATED_HUMAN_REVISED = "ai_generated_human_revised"
    HUMAN_AUTHORED = "human_authored"


class ClaimType(str, Enum):
    DETERMINISTIC_FACT = "deterministic_fact"
    INTERPRETATION = "interpretation"
    LIMITATION = "limitation"


class GroundingStatus(str, Enum):
    VERIFIED = "verified"
    UNSUPPORTED = "unsupported"
    UNVERIFIABLE = "unverifiable"


class AwardStatus(str, Enum):
    DRAFT = "draft"
    CONFIRMED = "confirmed"
    REVOKED = "revoked"


class AwardEventType(str, Enum):
    DRAFT_CREATED = "draft_created"
    CONFIRMED = "confirmed"
    REVOKED = "revoked"


# ---------------------------------------------------------------------------
# STRUCTURED FACT REFERENCE — Machine-readable context references
# ---------------------------------------------------------------------------


class FactReference(BaseModel):
    """Machine-readable reference to an authoritative DecisionContext value."""

    reference_type: str  # "supplier_score", "criterion_contribution", "rank", "normalized_value"
    supplier_id: str | None = None
    criterion_id: str | None = None
    field_path: str  # JSON path within DecisionContext, e.g. "suppliers[0].total_score"
    authoritative_value: str  # The exact string value from DecisionContext
    display_value: str | None = None  # Human-formatted display value


# ---------------------------------------------------------------------------
# NARRATIVE SECTIONS — Structured LLM output (instructor-enforced)
# ---------------------------------------------------------------------------


class SupplierAnalysis(BaseModel):
    """Per-supplier analysis section with structured fact references."""

    supplier_id: str
    supplier_name: str
    rank: int | None = None
    total_score: str  # Exact authoritative value rendered server-side
    strengths: list[str] = Field(default_factory=list)
    weaknesses: list[str] = Field(default_factory=list)
    score_context: str = ""  # Brief textual context about the score


class NarrativeSections(BaseModel):
    """
    Structured narrative output validated by instructor.
    Critical numeric values reference DecisionContext authoritatively.
    """

    executive_summary: str
    ranking_explanation: str
    per_supplier_analysis: list[SupplierAnalysis] = Field(default_factory=list)
    trade_offs: str = ""
    decision_considerations: str = ""
    risk_factors: list[str] = Field(default_factory=list)
    data_limitations: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# NARRATIVE CLAIM — Grounded claim representation
# ---------------------------------------------------------------------------


class NarrativeClaimSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    claim_index: int
    text: str
    claim_type: ClaimType
    grounding_status: GroundingStatus
    referenced_supplier_ids: list[str] = Field(default_factory=list)
    referenced_criterion_ids: list[str] = Field(default_factory=list)
    referenced_evidence_ids: list[str] = Field(default_factory=list)
    fact_references: list[dict[str, Any]] | dict[str, Any] | None = None
    grounding_notes: str | None = None


# ---------------------------------------------------------------------------
# NARRATIVE REVISION — Append-only human edit
# ---------------------------------------------------------------------------


class NarrativeRevisionCreate(BaseModel):
    revised_text: str = Field(..., min_length=1)
    revision_rationale: str | None = None


class NarrativeRevisionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    narrative_generation_id: str
    revision_number: int
    revised_text: str
    revision_rationale: str | None = None
    revised_by: str
    revised_at: datetime

    @field_validator("revised_by", mode="before")
    @classmethod
    def sanitize_revised_by(cls, v: Any) -> str:
        return _sanitize_actor_string(v)



# ---------------------------------------------------------------------------
# NARRATIVE GENERATION — Request / Response
# ---------------------------------------------------------------------------


class NarrativeGenerationRequest(BaseModel):
    scoring_run_id: str
    narrative_type: NarrativeType = NarrativeType.COMPARISON_SUMMARY
    include_sensitivity: bool = False
    human_unverified_note: str | None = None  # Explicitly unverified, never documentary evidence


class NarrativeGenerationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    rfq_id: str
    decision_context_id: str
    narrative_type: NarrativeType
    generation_number: int
    origin: NarrativeOrigin

    # LLM metadata
    provider: str
    model_identifier: str
    prompt_template_version: str
    prompt_template_hash: str
    rendered_prompt_hash: str
    response_schema_version: str
    generation_parameters: dict[str, Any]

    # Output
    raw_structured_output: dict[str, Any]
    output_hash: str
    grounding_validation_result: dict[str, Any]

    # Superseded tracking
    is_superseded: bool
    superseded_reason: str | None = None

    # Timestamps
    generated_at: datetime
    created_by: str

    @field_validator("created_by", mode="before")
    @classmethod
    def sanitize_created_by(cls, v: Any) -> str:
        return _sanitize_actor_string(v)

    # Derived — populated by service layer
    claims: list[NarrativeClaimSchema] = Field(default_factory=list)
    revisions: list[NarrativeRevisionResponse] = Field(default_factory=list)
    current_origin: NarrativeOrigin = NarrativeOrigin.AI_GENERATED


# ---------------------------------------------------------------------------
# DECISION CONTEXT — Response
# ---------------------------------------------------------------------------


class DecisionContextResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    rfq_id: str
    scoring_run_id: str
    context_schema_version: str
    context_payload: dict[str, Any]
    context_hash: str
    includes_sensitivity: bool
    created_by: str
    created_at: datetime

    @field_validator("created_by", mode="before")
    @classmethod
    def sanitize_created_by(cls, v: Any) -> str:
        return _sanitize_actor_string(v)


# ---------------------------------------------------------------------------
# AWARD DECISION — Lifecycle schemas
# ---------------------------------------------------------------------------


class AwardDecisionCreate(BaseModel):
    scoring_run_id: str
    narrative_generation_id: str | None = None
    awarded_supplier_id: str
    award_justification: str = Field(..., min_length=1)
    # Mandatory when awarded supplier is not Rank #1
    non_rank1_rationale: str | None = None


class AwardConfirm(BaseModel):
    final_justification: str | None = None  # Can override draft justification


class AwardRevoke(BaseModel):
    revocation_reason: str = Field(..., min_length=1)


class AwardDecisionEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    award_decision_id: str
    event_type: AwardEventType
    event_number: int
    event_payload: dict[str, Any]
    actor_principal: str
    actor_display_name: str | None = None
    created_at: datetime

    @field_validator("actor_principal", mode="before")
    @classmethod
    def sanitize_principal(cls, v: Any) -> str:
        return _sanitize_actor_string(v)



class AwardDecisionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    rfq_id: str
    scoring_run_id: str
    narrative_generation_id: str | None = None
    awarded_supplier_id: str
    awarded_supplier_name: str
    awarded_supplier_rank: int | None = None
    non_rank1_rationale: str | None = None
    current_status: AwardStatus
    provenance_hash: str | None = None
    created_by: str
    created_at: datetime
    events: list[AwardDecisionEventResponse] = Field(default_factory=list)

    @field_validator("created_by", mode="before")
    @classmethod
    def sanitize_created_by(cls, v: Any) -> str:
        return _sanitize_actor_string(v)

    # Derived — service layer populates
    is_based_on_latest_run: bool = True
    latest_scoring_run_id: str | None = None
